sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,652 @@ These follow the AAELF64_ base definition from Arm.
+is passed at least one, and at most 2 parameters dependent on the
+glibc IFUNC ABI version. A flag is set in the first parameter if a
+second parameter is available. The parameters provide access to the
+HWCAP_ flags.
+
+.. code-block:: c
+
+  /* A second argument is passed to the ifunc resolver.  */
+  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+
+  /* Function prototype for an IFUNC resolver.
+  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+
+  /* Second argument to an ifunc resolver.  */
+  struct __ifunc_arg_t
+  {
+    unsigned long _size; /* Size of the struct, so it can grow.  */
+    unsigned long _hwcap;
+    unsigned long _hwcap2;
+  };
+
+  typedef struct __ifunc_arg_t __ifunc_arg_t;
+
+IFUNC resolver functions must have a type of ``STT_GNU_IFUNC``. With
+the GCC and Clang compilers an attribute can be used to achieve this.
+
+.. code-block:: c
+
+  /* Implementation alternative 1 */
+  static int implementation1(void) { ... }
+
+  /* Implementation alternative 2 */
+  static int implementation2(void) { ... }
+
+  /* Resolver function */
+  static void* resolver(uint64_t, const __ifunc_arg_t *) { ... }
+
+  /* make symbol ifunc type STT_GNU_IFUNC using resolver as the IFUNC resolver. */
+  int ifunc(void) __attribute__((ifunc("resolver")));
+
+The first ``uint64_t`` parameter is ``AT_HWCAP``. If the
+``_IFUNC_ARG_HWCAP`` flag is set the second parameter is present. The
+second parameter is a structure ``__ifunc_arg_t`` defined in
+``sys/ifunc.h``, it provides access to all of the flags in HWCAP_ via
+fields of the structure.
+
+The IFUNC resolver function returns the address of the function
+implementation.
+
+IFUNC resolvers may be run when the dynamic linker is resolving relocations.
+additional restrictions on what they can contain.
+
+ * IFUNC resolvers must not be compiled with security features like
+   stack-protection, which requires a guard variable to be
+   initialized. Or instrumentation like ASAN that requires a shadow
+   map to be set up.
+
+ * IFUNC resolvers must not have a symbol binding of ``STB_WEAK``.
+
+The order of dynamic relocation resolution across an executable and
+all its shared libraries is platform specific. The following
+recommendations for writing IFUNC resolvers apply to the GNU glibc
+dynamic loader. Other dynamic linkers may have fewer requirements.
+
+ * IFUNC resolver functions must not call a function that may require
+   IFUNC initialization. If the IFUNC initialization for the called
+   function has not occured then undefined behavior results.
+
+ * In position-independent code an IFUNC resolver functions must not
+   call a function that requires a PLT entry. If the IFUNC resolver
+   runs as a result of a relocation in ``.rela.dyn`` then the
+   relocations in ``.rela.plt`` will not have been resolved. This means
+   that addresses in the ``.got.plt`` will be unchanged from their
+   static link time value.
```

Comment:
I'm a little confused about this.
My thought is that relocations in `.rela.plt` have a lazy mechanism and so why should this matter if the relocations are not yet resolved.
Not sure if that makes sense in the case of `BIND_NOW` -- not sure if there are any other assumptions in that case -- is that the reason why this matters?
Or is it something to do with some other setup of the PLT mechanism in the dynamic linker?

Reply:
In a position independent executable or shared library all the addresses in the .got.plt need to have the load bias added to them at initialisation time. This needs to be done even when lazy resolving as the address of the PLT entry for the lazy resolver in the .got.plt needs to be adjusted by the load bias.

FWIW this confused me too until I tried to write a program to see what happened.